Fix automatic min size handling of flex and grid items to match the specification

Ahmad-S792 opened the pull request:
#### 6ec233675a708b3ac9d0cc463f1dfbb753119501
<pre>
Fix automatic min size handling of flex and grid items to match the specification
<a href="https://bugs.webkit.org/show_bug.cgi?id=297325">https://bugs.webkit.org/show_bug.cgi?id=297325</a>
<a href="https://rdar.apple.com/158215991">rdar://158215991</a>

Reviewed by Sammy Gill.

This patch aligns WebKit with Gecko / Firefox and Blink / Chromium.

Inspired by: <a href="https://chromium-review.googlesource.com/c/chromium/src/+/5505559">https://chromium-review.googlesource.com/c/chromium/src/+/5505559</a>

Both the grid and flex specs mention &quot;scroll container&quot; already, which
means that visible and clip should behave the same:

 * <a href="https://drafts.csswg.org/css-grid/#min-size-auto">https://drafts.csswg.org/css-grid/#min-size-auto</a>
 * <a href="https://drafts.csswg.org/css-flexbox/#min-size-auto">https://drafts.csswg.org/css-flexbox/#min-size-auto</a>

This patch adds `Overflow: Clip` handling relevant functions.

* Source/WebCore/rendering/GridTrackSizingAlgorithm.cpp:
(WebCore::GridTrackSizingAlgorithmStrategy::minContributionForGridItem const):
* Source/WebCore/rendering/RenderFlexibleBox.cpp:
(WebCore::RenderFlexibleBox::shouldApplyMinSizeAutoForFlexItem const):
* LayoutTests/TestExpectations: Progressions

Canonical link: <a href="https://commits.webkit.org/298690@main">https://commits.webkit.org/298690@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/726bab13dc87a38d0ec29c247a6454df3e0f5abf

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/116363 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/36024 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/26574 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/122420 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/66924 "Built successfully") | ⏳ 🛠 ios-apple 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/118252 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/36718 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/44612 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/88382 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/42861 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | ⏳ 🛠 mac-apple 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/119312 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/29281 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/104396 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/68815 "Passed tests") | | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/28363 "Passed tests") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/135/builds/22499 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/66101 "Built successfully") | | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/98663 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/22660 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/125569 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/43257 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/32484 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/97089 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/43622 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/100602 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/96884 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/24650 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/42172 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/20071 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/39211 "Built successfully") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/43145 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/48736 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/42611 "Built successfully") | | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/45951 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/44316 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

sammygill commented on Source/WebCore/rendering/GridTrackSizingAlgorithm.cpp:
```diff
@@ -1168,7 +1168,7 @@ LayoutUnit GridTrackSizingAlgorithmStrategy::minContributionForGridItem(RenderBo
 
     auto& gridItemMinSize = isRowAxis ? gridItem.style().logicalMinWidth() : gridItem.style().logicalMinHeight();
-    bool overflowIsVisible = isRowAxis ? gridItem.effectiveOverflowInlineDirection() == Overflow::Visible : gridItem.effectiveOverflowBlockDirection() == Overflow::Visible;
+    bool overflowIsVisible = isRowAxis ? (gridItem.effectiveOverflowInlineDirection() == Overflow::Visible || gridItem.effectiveOverflowInlineDirection() == Overflow::Clip) : gridItem.effectiveOverflowBlockDirection() == Overflow::Visible;
```
I think it would be nice if we didn't have to potentially call `effectiveOverflowInlineDirection()` twice so maybe we should store the overflow value in a local variable here. 

Also `overflowIsVisible` should probably be renamed since that's not what we're looking for anymore. The spec says "its computed overflow is not a scrollable overflow value" so we should probably try to make it match in some way

Ahmad-S792 replied:
@sammygill - does it look good now?

sammygill commented on Source/WebCore/rendering/RenderFlexibleBox.cpp:
```diff
@@ -575,7 +575,7 @@ bool RenderFlexibleBox::shouldApplyMinSizeAutoForFlexItem(const RenderBox& flexI
     bool flexItemBlockSizeIsEquivalentToAutomaticSize  = !mainAxisIsFlexItemInlineAxis(flexItem) && (minSize.isMinContent() || minSize.isMaxContent() || minSize.isFitContent());
 
-    return (minSize.isAuto() || flexItemBlockSizeIsEquivalentToAutomaticSize) && (mainAxisOverflowForFlexItem(flexItem) == Overflow::Visible);
+    return (minSize.isAuto() || flexItemBlockSizeIsEquivalentToAutomaticSize) && (mainAxisOverflowForFlexItem(flexItem) == Overflow::Visible || mainAxisOverflowForFlexItem(flexItem) == Overflow::Clip);
```
I think it would be nice if we didn't have to potentially call `mainAxisOverflowForFlexItem` twice. Maybe we could factor this out into a lambda or separate function and call it when needed

sammygill commented on Source/WebCore/rendering/GridTrackSizingAlgorithm.cpp:
```diff
@@ -1168,11 +1168,16 @@ LayoutUnit GridTrackSizingAlgorithmStrategy::minContributionForGridItem(RenderBo
 
     if (gridItemMinSize.isAuto()) {
-        if (!overflowIsVisible)
+        if (!overflowIsNotScrollable())
```
This reads "if not overflow is not scrollable," which I personally am not the biggest fan of due to the double negative, but it's probably fine either way.

Ahmad-S792 committed:
Fix automatic min size handling of flex and grid items to match the specification
https://bugs.webkit.org/show_bug.cgi?id=297325
rdar://158215991

Reviewed by Sammy Gill.

This patch aligns WebKit with Gecko / Firefox and Blink / Chromium.

Inspired by: https://chromium-review.googlesource.com/c/chromium/src/+/5505559

Both the grid and flex specs mention "scroll container" already, which
means that visible and clip should behave the same:

 * https://drafts.csswg.org/css-grid/#min-size-auto
 * https://drafts.csswg.org/css-flexbox/#min-size-auto

This patch adds `Overflow: Clip` handling relevant functions.

* Source/WebCore/rendering/GridTrackSizingAlgorithm.cpp:
(WebCore::GridTrackSizingAlgorithmStrategy::minContributionForGridItem const):
* Source/WebCore/rendering/RenderFlexibleBox.cpp:
(WebCore::RenderFlexibleBox::shouldApplyMinSizeAutoForFlexItem const):
* LayoutTests/TestExpectations: Progressions

Canonical link: https://commits.webkit.org/298690@main
